todo-list/src/components/ListItem.vue

```diff
@@ -0,0 +1,80 @@
+<template>
+    <div class="row">
+        <div class="column">
+            <input type="checkbox" name="done" v-model="todo.done">
+        </div>
+        <div class="column">
+            <input v-if="editMode" v-model="tempName"
+                   @keyup.enter="saveNewName" @keydown.esc="cancelNewName" />
+            <p v-else :class="{ 'completed-todo-item': todo.done}" >{{todo.name}}</p>
+        </div>
+        <div v-if="editMode" class="column">
+            <button @click="saveNewName">Save</button>
+            <button @click="cancelNewName">Cancel</button>
+        </div>
+        <div v-else class="column">
+            <button @click="startEditMode">Edit</button>
+            <button @click="onDeleteClicked">Delete</button>
+        </div>
+    </div>
+</template>
+
+<script lang="ts">
+    import { Component, Prop, Vue } from "vue-property-decorator";
+    import {ToDo} from "@/todo";
+
+    @Component
+    export default class ListItem extends Vue {
+        @Prop() private todo!: ToDo;
+        @Prop() private id!: number;
+
+        private editMode: boolean = false;
+        private tempName: string = "";
+
+        startEditMode() {
+            this.tempName = this.todo.name;
+            this.editMode = true;
+        }
+
+        saveNewName() {
+            if (!this.editMode) {
+                return;
+            }
+
+            this.$emit("on-name-changed", this.todo.id, this.tempName);
```

Comment:
~Is it possible to emit `this.todo` here? The fact that you added `this.tempName` looks like it might be necessary. Is it necessary?~ :thinking: 
Do you get a warning that properties should not be edited in-place when you set `this.todo.message = tempName` and emit the whole `todo`?

Reply:
yes exactly we get a warning, that we should not change the vsriables set by vue.js